### ISSUE TITLE
Can I put the edges' weights into the edge_attr?

### ISSUE THREAD

**Issue by gmftbyGMFTBY** [pyg-team/pytorch_geometric#708]
## ❓ Questions & Help

Hi, first of all, thank you so much for this wonderful tool for PyTorch.
I am a newbie in this library so I have two very strange questions.

---
### 1. `edge_attr`
I notice that the `Data` class has an attribution named `edge_attr` which holds the edges' features in the graph.
And currently, I need to construct a weighted directed graph `Data` instance for `GCNConv`.
So here is my question: Can I put the weight into the `edge_attr` for creating the `Data` instance?
In my view, I think it is just okay, but seems so strange. Because in this time the shape of the `edge_attr` tensor is actually `[num_edges, 1]`.

So, is it okay to do it?

---

### 2. `Dynamical batch`
I also want to confirm one thing.
Here [FAQ2](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_dataset.html), your introduction seems to show that I can dynamically construct a graph `DataLoader` in this way:

```python
from torch_geometric.data import Data, DataLoader

data_list = [Data(...), ..., Data(...)]
loader = DataLoader(data_list, batch_size=32)
```
Is that all I need to do? Don't I need to do some stuff like padding for the batches?

Finally, thank you so much for your work and hope you to respond to my questions.

**Comment by gmftbyGMFTBY** [comment]
As shown in issue #707 , in order to process the weighted graph in mini-batch settings, Can I merge all the sub `edge_weight` tensors to a **long** big tensor which will be used by the mini-batch graph?

**Comment by rusty1s** [comment]
You can save your edge weights into `edge_attr`. This is no problem. Alternatively, you can save them to any attribute, *e.g.*, `data.edge_weight = edge_weight`. The `DataLoader` correctly merges all attributes into a single graph. So this is indeed all you need to do.

**Comment by gmftbyGMFTBY** [comment]
Thank you for your response. 
After reading your introductions, I know that you merge all the graph in the mini-batch into a single big graph. 
So in order to apply the `GCNConv` to process the weighted graph, I can also `torch.cat` all the `edge_weigth` into a long `edge_weight` for the merged single graph which is shown in #707. 
Am I right?

Thank you so much!

**Comment by dabao0903** [comment]
> You can save your edge weights into `edge_attr`. This is no problem. Alternatively, you can save them to any attribute, _e.g._, `data.edge_weight = edge_weight`. The `DataLoader` correctly merges all attributes into a single graph. So this is indeed all you need to do.

Hello, author. Can you explain the principle of "Weighted GCN layer", thanks

**Comment by rusty1s** [comment]
We follow the original paper here and compute the degree matrix based on edge weights. Hence, normalization is both dependent on edge weights and node degrees.

**Comment by udithhaputhanthri** [comment]
@rusty1s 
Hello author, 
Thank you for the tool and it is really cool.

I would like to know that is there a way to implement the "Graph Attention Network" model (GAT) to train the edge weights of the graph. If so, it is great if you can mention the documentation/ any other document.  Or can it be achieved by changing neighborhood aggregation function in the GCN layer to something like "pool"? 

Please reply

**Comment by rusty1s** [comment]
Please have a look at the GATConv layer in torch_geometric.nn.
